src/OpenSage.Game/Data/Apt/Geometry.cs

```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
+                                    if (paramList.Length != 12)
+                                        throw new InvalidDataException();
+
+                                    style = GeometryStyle.TexturedTri;
+                                    color.R = Convert.ToByte(paramList[1]);
+                                    color.G = Convert.ToByte(paramList[2]);
+                                    color.B = Convert.ToByte(paramList[3]);
+                                    color.A = Convert.ToByte(paramList[4]);
+                                    //image id used
+                                    image = Convert.ToInt32(paramList[5]);
+                                    //transformation parameters, to map the geometry above the texture
+                                    rotMat.M11 = ParseUtility.ParseFloat(paramList[6]);
+                                    rotMat.M12 = ParseUtility.ParseFloat(paramList[7]);
+                                    rotMat.M21 = ParseUtility.ParseFloat(paramList[8]);
+                                    rotMat.M22 = ParseUtility.ParseFloat(paramList[9]);
+                                    translation.X = ParseUtility.ParseFloat(paramList[10]);
+                                    translation.Y = ParseUtility.ParseFloat(paramList[11]);
+                                    break;
+                            }
+                            break;
+                        //A line
+                        case 'l':
+                            Line2D lin;
+                            lin.V0 = new Vector2(ParseUtility.ParseFloat(paramList[0]), ParseUtility.ParseFloat(paramList[1]));
+                            lin.V1 = new Vector2(ParseUtility.ParseFloat(paramList[2]), ParseUtility.ParseFloat(paramList[3]));
+                            lines.Add(lin);
+                            break;
+                        //A triangle
+                        case 't':
+                            Triangle2D tri;
+                            tri.V0 = new Vector2(ParseUtility.ParseFloat(paramList[0]), ParseUtility.ParseFloat(paramList[1]));
+                            tri.V1 = new Vector2(ParseUtility.ParseFloat(paramList[2]), ParseUtility.ParseFloat(paramList[3]));
+                            tri.V2 = new Vector2(ParseUtility.ParseFloat(paramList[4]), ParseUtility.ParseFloat(paramList[5]));
+                            tris.Add(tri);
+                            break;
+                    }
+
                 }
+
+                ApplyStyle();
             }
 
+            geometry.CalculateBoundings();
             return geometry;
         }
+
+        private void CalculateBoundings()
+        {
+            //find topleft and botright point
+            var topLeft = new Vector2 { X = float.MaxValue, Y = float.MaxValue };
+            var botRight = new Vector2 { X = float.MinValue, Y = float.MinValue };
+
+            foreach (var entry in Entries)
+            {
+                if (entry is GeometryLines gl)
```

Comment:
Nice use of C#'s new pattern syntax :) I think you could go one step further and use the new switch/case pattern matching:

``` csharp
switch (entry)
{
  case GeometryLines gl:
    break;

  case GeometrySolidTriangles gst:
    break;

  // etc.
}
```

Reply:
Wow, that's really amazing 👍